mypy/checker.py

```diff
@@ -445,8 +446,22 @@ def _visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
             self.check_func_item(fdef.func, name=fdef.func.name())
             if fdef.func.is_abstract:
                 num_abstract += 1
+            if fdef.func.is_awaitable_coroutine:
+                num_awaitable_coroutine += 1
         if num_abstract not in (0, len(defn.items)):
             self.fail(message_registry.INCONSISTENT_ABSTRACT_OVERLOAD, defn)
+        if num_awaitable_coroutine not in (0, len(defn.items)):
+            self.fail(message_registry.INCONSISTENT_COROUTINE_OVERLOAD, defn)
+        # If items contains coroutines and check_func_item fixed their type,
+        # also fix the overload type.
+        if num_awaitable_coroutine:
+            assert isinstance(defn.type, Overloaded)
+            types = []
+            for fdef, typ in zip(defn.items, defn.type.items()):
+                assert isinstance(fdef, Decorator)
```

Comment:
I wonder if we could get a crash here if one of the overloads doesn't have `@coroutine`.

Reply:
Yeah -- maybe we could just skip this entire block if `@coroutine` is applied inconsistently? It's probably not worth trying to infer more precise types once we know the decorators are wonky.